### PR TITLE
Kafka registry

### PR DESCRIPTION
**Pull request by khevse**

### CODE REVIEW

File: kafka/registry/registry.go [hunk 1]
```diff
@@ -0,0 +1,347 @@
+package registry
```
Comment (Badmoonz) [inline kafka/registry/registry.go]: rename to schema_registry

File: kafka/registry/registry.go [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+
+	sendBodyErr := make(chan error, 1)
+	{
+		// prepare http request
+		req.Header = header
+		req = req.WithContext(ctx)
+
+		if payload == nil {
+			sendBodyErr <- nil
+			defer close(sendBodyErr)
```
Comment (Badmoonz) [inline kafka/registry/registry.go]: why defer?

File: kafka/registry/registry.go [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+	}
+
+	res, err := r.client.Do(req)
+	if err == nil {
+		err = <-sendBodyErr
+		defer res.Body.Close()
+	}
+
+	if err != nil {
+		res.Body.Close()
```
Comment (Badmoonz) [inline kafka/registry/registry.go]: res.Body can't be nil if err comes from client.Do ?

File: kafka/registry/registry.go [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+
+			go func() {
+				defer reqBodyWriter.Close()
+				defer close(sendBodyErr)
+
+				sendBodyErr <- json.NewEncoder(reqBodyWriter).Encode(payload)
+			}()
+
+			req.Body = reqBodyReader
+			req.GetBody = func() (io.ReadCloser, error) {
```
Comment (Badmoonz) [inline kafka/registry/registry.go]: isn't Request.GetBody defined by default?

File: kafka/registry/registry.go [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+)
+
+const ContentType = "application/vnd.schemaregistry+json"
+
+type Registry struct {
+	client  *http.Client
+	uriPool sync.Pool
+}
+
+func New(uri string, timeout time.Duration, certPem []byte) (*Registry, error) {
```
Comment (Badmoonz) [inline kafka/registry/registry.go]: certPem better to abstract over http.Transport

Comment (Badmoonz) [reply kafka/registry/registry.go]: add option for BasicAuth

File: Makefile [hunk 6]
```diff
@@ -3,9 +3,10 @@
 PACKAGES_WITH_TESTS:=$(shell go list -f="{{if or .TestGoFiles .XTestGoFiles}}{{.ImportPath}}{{end}}" ./... | grep -v '/vendor/')
 TEST_TARGETS:=$(foreach p,${PACKAGES_WITH_TESTS},test-$(p))
 TEST_OUT_DIR:=testout
+PROJECT:=github.com/dialogs/dialog-go-lib
 
 .PHONY: all
-all: mod static mock proto lint testall
+all: static mock proto mod easyjson mod lint testall
```
Comment (alexboyko) [inline Makefile]: mod + mod?

Comment (khevse) [reply Makefile]: It's not a error. Easyjson can't generate code without dependencies. And after, need to add new dependency for easyjson.

File: kafka/schemaregistry/client.go [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+	u.RawQuery = ""
+	u.Fragment = ""
+
+	c.uriPool.Put(u)
+}
+
+// getID returns id for url
+func getID(src int) (id string) {
+
+	if src == -1 {
```
Comment (alexboyko) [inline kafka/schemaregistry/client.go]: I think magic -1 number should be mentioned in doc

**Pull request merged**